### ISSUE TITLE
ActiveRecord::Enum and inheritance bug

### ISSUE THREAD

**Issue by thepieterdc** [rails/rails#39756]
### Steps to reproduce
I have uploaded a very minimal working example to https://github.com/thepieterdc/rails-bug-mwe; the README contains instructions on how to reproduce the bug.

When using enums on a class that uses inheritance, the helper methods do not work correctly and the enum constants are not translated from integers to strings and back, resulting in loss of information. I have not yet pinned the exact cause of the failure, but after some searching around, it might somehow be related to https://github.com/rails/rails/issues/16288

### Expected behavior
(based on the test case in https://github.com/thepieterdc/rails-bug-mwe/blob/master/test/provider_test.rb)
`prov.prefer!` should set the `mode` enum in the database, after which `prov.prefer?` should return true, adhering to the documentation of ActiveRecord enums.

### Actual behavior
(based on the test case in https://github.com/thepieterdc/rails-bug-mwe/blob/master/test/provider_test.rb)
`prov.prefer!` does not modify the database and therefore `prov.prefer?` returns false, failing the assertion and therefore the test case.

```
Run options: --seed 51653

# Running:

F

Failure:
ProviderTest#test_demo [/projects/bug/test/provider_test.rb:7]:
Expected false to be truthy.

rails test test/provider_test.rb:4

Finished in 0.129250s, 7.7370 runs/s, 7.7370 assertions/s.
1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

### System configuration
**Rails version**: 6.0.3

**Ruby version**: 2.6.5

**Comment by eugeneius** [comment]
Thanks for the report! The problem here is that the subclasses are being defined before the enum attribute is added to the base class, and attributes are not currently copied onto subclasses after the fact. This is set to be fixed in https://github.com/rails/rails/pull/39468 or https://github.com/rails/rails/pull/39487.

If you need a workaround for now, you could reorder the base class so that the enum attribute is defined first:

```patch
diff --git a/app/models/provider.rb b/app/models/provider.rb
index 1fdfd25..34b85e6 100644
--- a/app/models/provider.rb
+++ b/app/models/provider.rb
@@ -1,5 +1,5 @@
 class Provider < ActiveRecord::Base
-  PROVIDERS = [Provider::GSuite, Provider::Office365].freeze
-
   enum mode: { prefer: 0, redirect: 1 }
+
+  PROVIDERS = [Provider::GSuite, Provider::Office365].freeze
 end
```

**Comment by thepieterdc** [comment]
> Thanks for the report! The problem here is that the subclasses are being defined before the enum attribute is added to the base class, and attributes are not currently copied onto subclasses after the fact. This is set to be fixed in #39468 or #39487.
> 
> If you need a workaround for now, you could reorder the base class so that the enum attribute is defined first:
> 
> ```diff
> diff --git a/app/models/provider.rb b/app/models/provider.rb
> index 1fdfd25..34b85e6 100644
> --- a/app/models/provider.rb
> +++ b/app/models/provider.rb
> @@ -1,5 +1,5 @@
>  class Provider < ActiveRecord::Base
> -  PROVIDERS = [Provider::GSuite, Provider::Office365].freeze
> -
>    enum mode: { prefer: 0, redirect: 1 }
> +
> +  PROVIDERS = [Provider::GSuite, Provider::Office365].freeze
>  end
> ```

This actually makes a lot of sense, thanks alot for your fast and quick reply, including the workaround!
